walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,93 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+
 
 public final class FindMeetingQuery {
+
+  // create a set of events to avoid duplicates
+  private Collection<Event> events = new HashSet<>();
+
+  // the meeting request
+  private MeetingRequest request; 
+
+  private static final Collection<Event> NO_EVENTS = Collections.emptySet();
+  private static final Collection<String> NO_ATTENDEES = Collections.emptySet();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+ 
+    this.events = events;
+    this.request = request; 
+
+    // Check if duration is longer than a day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+        return Arrays.asList(); 
+    }
+    // Check for no events or no attendees
+    if (events == NO_EVENTS || request.getAttendees() == NO_ATTENDEES) {
+        return Arrays.asList((TimeRange.WHOLE_DAY)); 
+    }
+    // Arraylist holds event times accordingly     
+    List<TimeRange> mandatory = new ArrayList();
+
+    for (Event event: events) {
+
+        // Check if attendees from event are requested for meeting
+        for (String attendee : event.getAttendees()) {            
+            // Check that attendee is a mandatory attendee
+            if ((request.getAttendees().contains(attendee)) ) {
+                mandatory.add(event.getWhen());
+            }           
+        }
+    }
```

Comment:
consider using Java 8's stream() api. Something like this (untested):
```suggestion
events.stream().filter(
    event.getAttendees().stream().anyMatch(
        a -> request.getAttendees().contains(attendee))
    .map(e -> e.getWhen()).collect(Collectors.ToList());
```